compiler/src/iree/compiler/InputConversion/StableHLO/Preprocessing/StableHLOToStableHLO.cpp

```diff
@@ -1540,6 +1540,71 @@ struct CustomCallIsTopK final
   }
 };
 
+struct IotaSortSliceIsTopK final : OpRewritePattern<mlir::stablehlo::SortOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(mlir::stablehlo::SortOp op,
+                                PatternRewriter &rewriter) const override {
+    auto opOperands = op.getOperands();
+    auto opResults = op.getResults();
+    Value topKInput;
+    if (opOperands.size() != 2 || opResults.size() != 2) {
+      return rewriter.notifyMatchFailure(
+          op, "Slice that maps to TopK must have exactly two inputs/outputs");
+    }
+
+    bool isInputIota = false;
+    // Check that one of the inputs is iota, assume that the other one is the
+    // input.
+    for (Value operand : opOperands) {
+      auto iotaOp =
+          dyn_cast_or_null<mlir::stablehlo::IotaOp>(operand.getDefiningOp());
+      if (iotaOp) {
+        isInputIota = true;
+      } else {
+        topKInput = operand;
+      }
+    }
+
+    if (!isInputIota) {
+      return rewriter.notifyMatchFailure(op, "Sort isn't called from Iota.");
+    }
+
+    Value topV, topI;
+    int64_t k;
+    // Check that the output of the sort op gets fed into a slice.
+    for (auto [idx, result] : llvm::enumerate(opResults)) {
+      auto sliceOp =
+          dyn_cast<mlir::stablehlo::SliceOp>(*result.getUsers().begin());
+      if (!sliceOp) {
+        return rewriter.notifyMatchFailure(
+            op, "Sort isn't calling into a slice op.");
+      }
+
+      for (auto stride : sliceOp.getStrides().getValues<int64_t>()) {
+        if (stride != 1) {
+          return rewriter.notifyMatchFailure(
+              op, "All slice strides must be 1 in order to match to TopK.");
+        }
+      }
+
+      // Treat the first slice as inputs, the second as indices.
+      if (idx == 0) {
+        topV = sliceOp.getResult();
+        k = sliceOp.getLimitIndices().getValues<int64_t>()[1];
+      } else {
+        topI = sliceOp.getResult();
+      }
+    }
+
+    auto topK = rewriter.create<chlo::TopKOp>(
+        op.getLoc(), TypeRange{topV.getType(), topI.getType()}, topKInput, k);
```

Comment:
I have a meta question, what should we do if top-k is actually not going to be performant?
Do we always match top-k at this level?

Then the backend would have to undo this transformation / do the right thing if top-k is not the right op in this case (e.g., when K is close to the full size).

Reply:
That is a good questions. Specific performance requirements would probably be handled per backend. E.g. if we see a top-k has k = N, we would just run a sorting algo. This would also be true for k ~= N. Those kinds of choices would need to be backend specific though.

Reply:
Or phrased another way - even without this matcher we could encounter a `stablehlo.topk` with this case. Because `stablehlo.topk` is a more specialized case it would make sense we take the fastest path (whether pattern matched to `topk` or directly invoked).

Reply:
Good point!